fix #19373 move all hairpins, when inserting measure into part

sammik opened the pull request:
Resolves: #19373
Resolves: #19587
Resolves partialy (second issue) of #19476

<!-- Add a short description of and motivation for the changes here -->
~~undoInsertTime "acts on the linked scores as well", but only in score, there are all elements to be linked~~
inserting Measures always acts in `masterScore` and adds measures to all scores
inserting Frames acts in local score


<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

sammik commented:
this is somehow similar as #19271

question is, if `undoInsertTime` should also be always called from master score, or in other cases, it is good to call it from "local" score

@cbjeukendrup ?

cbjeukendrup commented:
Yes, it would make sense to me to perform all operations relating to inserting/removing measures via the MasterScore, so inserting and deleting measures too. 
Same for the "remove trailing empty measures" command: #17350

sammik commented:
> Yes, it would make sense to me to perform all operations relating to inserting/removing measures via the MasterScore, just like in #19271. 

It is not possible, as `insertMeasureOtions` have property `addToAllScores`, so in some situations, it is performed on just a particular excerpt score.

I can imagine something like
```
Score::insertMeasure(ElementType type, MeasureBase* beforeMeasure, const InsertMeasureOptions& options)
{
    if (options.addToAllScores && !score().isMaster()) {
        ...
        masterScore().insertMeasures(insertMeasure(type, masterBeforeMeasure, options);
        ...
        return ...;
    }
    ...
}
```
Not sure, if it is good idea.
@cbjeukendrup

sammik commented:
> Perhaps we should split the method into two methods: one for the elementary operation of adding measures to one `Score` instance only, and one for the score editing operation of adding measures globally. The elementary local operation should be in the Score class; the score editing operation should be in the MasterScore class. The global operation should probably call the local operation for all scores from `scoresList()`. You might need to add another method to `Score` that is a wrapper around the global method, for example to facilitate selecting the newly inserted measures in the current score.
> 
> What do you think about that idea?

It seems to me a bit overcomplicated, and somehow I dont see many benefits over [my current implementation](https://github.com/musescore/MuseScore/pull/19405/files), but sure, I can try.

sammik commented:
A little bit a different solution.

I found some problems with Your propose - there are bits of code, which would make sense to be called just once, in both cases - after we do some job on just local score / multiple scores

It would require to split whole method to multiple fragments.

So I did it this way:
- splitMeasure is acting always on masterScore (in local score there is wrapper to get local inserted measure)
- I removed option `addToAllScores` as it has to be always true now
- in only single case, where `addToAllScores = false` (only local score) were used: creating header frame for excerpts on scores without header frame on mastrer score I changed to be without `insertMeasure` method

@cbjeukendrup

cbjeukendrup commented on src/engraving/dom/excerpt.cpp:
```diff
@@ -347,10 +347,12 @@ void Excerpt::createExcerpt(Excerpt* excerpt)
     if (!scoreMeasure || !scoreMeasure->isVBox()) {
-        Score::InsertMeasureOptions options;
-        options.addToAllScores = false;
+        MeasureBase* newMeasureBase = toMeasureBase(Factory::createItem(ElementType::VBOX, score->dummy()));
```
```suggestion
        MeasureBase* newMeasureBase = Factory::createVBox(score->dummy()->system());
```

cbjeukendrup commented:
@sammik That seems a good solution! But now a rebase is needed, to include the changes we made to `insertMeasure` in that other PR...

sammik commented:
@cbjeukendrup done!

sammik commented:
@cbjeukendrup 
it should be finished now

I am sorry, but I needed to change many things, so it requires another proper review

I split original `insertMeasure` into 
- `MasterScore::insertMeasure` which is used to inserting measures, and adds new measure to all scores
- `Score::insertBox` which is used to insert frames, and adds frame to single score only

cbjeukendrup commented on src/engraving/dom/box.cpp:
```diff
@@ -431,7 +431,7 @@ void Box::manageExclusionFromParts(bool exclude)
             MeasureBase* newMB = nullptr;
-            if (!excludeFromOtherParts()) {
+            if (!next()->isMeasure() && !next()->excludeFromOtherParts()) {
                 for (auto e : linkList()) {
```
Shouldn't this be `next()->linkList()` as well? (assuming that `newMB` is supposed to be the new `next()` for the new box that will be created)

sammik replied:
Yes, of course, thanks.

sammik replied:
@cbjeukendrup I was also thinking, if not to create new function `MeasureBase::getInScore(Score*, bool useNextMeasureFallback")` which returns MeasureBase of given score.

Than I could replace all that new code by `MeasureBase* newMB = next().getInScore(score, true);`

What do You think?

cbjeukendrup replied:
That sounds like a good idea!

cbjeukendrup commented on src/engraving/dom/excerpt.cpp:
```diff
@@ -348,10 +348,7 @@ void Excerpt::createExcerpt(Excerpt* excerpt)
-        options.addToAllScores = false;
-
-        score->insertMeasure(ElementType::VBOX, scoreMeasure, options);
+        score->insertMeasure(ElementType::VBOX, scoreMeasure);
```
I would prefer to use `insertBox` in those places where the type is explicitly a box type anyway

cbjeukendrup commented:
Apart from these last two comments it looks good to me now!

sammik commented:
> Apart from these last two comments it looks good to me now!

@cbjeukendrup resolved and rebased.

All changes are in new commits

oktophonie commented:
With this PR I can't get 'Insert [bar] before selection' to do anything at all.

sammik commented:
> With this PR I can't get 'Insert [bar] before selection' to do anything at all.

Thx, Ill check it tonight.

sammik commented:
@oktophonie done
